File: preprocessor/hasinclude.go
package preprocessor

import "github.com/vertex-language/vcc/token"

// __has_include, and the operators of its shape.
//
// A header that wants a header it may not have has no other way to ask.
// Before this existed the question was answered by the build system --
// configure ran a compile and wrote a macro -- and the SDK on every Mac
// stopped doing that: Availability.h asks
//
//	#if __has_include(<AvailabilityInternalPrivate.h>)
//
// on line 199, so a compiler without the operator cannot read <stdlib.h>,
// and therefore cannot read anything.
//
// It is an operator rather than a macro, and for the same reason `defined`
// is one: its operand is not an expression. `__has_include(<sys/types.h>)`
// has a header-name inside it, which phase 3 does not produce and macro
// expansion must not touch -- expanding it would rewrite `sys`, and the
// slash and the dot were never tokens at all. So it is resolved before
// expansion, over the line as written, exactly where `defined` is.
//
// The answer is whether an #include written here would find a file. Not
// whether one exists somewhere: the search list and the including file's own
// directory are what decide, which is why this asks the same searchList the
// directive asks. An operator that answered a different question from the
// directive it guards would be worse than no operator.

const (
	hasInclude     = "__has_include"
	hasIncludeNext = "__has_include_next"
)

// builtinPPMacro reports whether a name is one of the operators the
// preprocessor answers for itself.
//
// `defined(__has_include)` and `#ifdef __has_include` are how a portable
// header asks whether it may use the operator, so both have to say yes --
// and neither may be answered by defining a macro of that name, because a
// macro would be expanded and the operand would be expanded with it.
func builtinPPMacro(name string) bool {
	return name == hasInclude || name == hasIncludeNext
}

// resolveHasInclude replaces every `__has_include(...)` in a controlling
// expression with 1 or 0.
//
// Run after resolveDefined and before expansion: `defined(__has_include)` is
// a question about the operator and is already a number by the time this
// looks, and what is left is the operator applied to a header-name.
func (p *Preprocessor) resolveHasInclude(r *reader, line []Token, at Site) []Token {
	out := make([]Token, 0, len(line))
	for i := 0; i < len(line); i++ {
		t := line[i]
		if !t.Is(hasInclude) && !t.Is(hasIncludeNext) {
			out = append(out, t)
			continue
		}
		next := t.Is(hasIncludeNext)
		what := hasInclude
		if next {
			what = hasIncludeNext
		}

		if i+1 >= len(line) || line[i+1].Kind != token.LPAREN {
			p.errorf(t.Site(), "operator \"%s\" requires a header name in parentheses", what)
			out = append(out, p.number(t, 0))
			continue
		}
		close, ok := matchParen(line, i+1)
		if !ok {
			p.errorf(t.Site(), "missing ')' after \"%s\"", what)
			out = append(out, p.number(t, 0))
			return append(out, line[i+1:]...)
		}

		// The operand alone, so headerName sees a whole line and its
		// end-of-line check is about the parentheses rather than the rest of
		// the expression.
		name, angled, found := p.headerName(what, line[i+1+1:close], t.Site())
		n := 0
		if found && p.headerExists(r, name, angled, next) {
			n = 1
		}
		out = append(out, p.number(t, n))
		i = close
	}
	return out
}

// matchParen is the index of the ')' closing the '(' at open, and reports
// whether there was one. Nesting is counted because a header name has no
// parentheses but a macro that expands to one may.
func matchParen(line []Token, open int) (int, bool) {
	depth := 0
	for i := open; i < len(line); i++ {
		switch line[i].Kind {
		case token.LPAREN:
			depth++
		case token.RPAREN:
			depth--
			if depth == 0 {
				return i, true
			}
		}
	}
	return 0, false
}

File: tests/preproc/has_include.c
#include <stdio.h>

/* __has_include, which is an operator rather than a macro: its operand is a
   header-name, and a header-name is not an expression. Every check here is
   one a real system header makes -- the SDK on a Mac cannot be read at all
   without this, because Availability.h asks before it declares anything. */

/* The operator answers `defined`, which is how a portable header finds out
   whether it may use it. */
#if !defined(__has_include)
#error "__has_include should be defined"
#endif

#ifndef __has_include
#error "__has_include should answer #ifdef too"
#endif

/* A header that is there, and one that is not. */
#if __has_include(<stdio.h>)
#define HAVE_STDIO 1
#else
#define HAVE_STDIO 0
#endif

#if __has_include(<no/such/header.h>)
#define HAVE_BOGUS 1
#else
#define HAVE_BOGUS 0
#endif

/* The quoted form looks beside this file first, which is the whole
   difference between the two spellings. */
#if __has_include("has_a/present.h")
#define HAVE_LOCAL 1
#else
#define HAVE_LOCAL 0
#endif

#if __has_include("has_a/absent.h")
#define HAVE_LOCAL_BOGUS 1
#else
#define HAVE_LOCAL_BOGUS 0
#endif

/* Asking does not include: the macro that header defines is not defined
   here. */
#ifdef PRESENT_H_WAS_READ
#error "__has_include read the header instead of looking for it"
#endif

/* It composes with the rest of a controlling expression, and the operand
   with a slash and a dot in it survives -- neither is a token between the
   angle brackets. */
#if defined(__has_include) && __has_include(<sys/types.h>) && !__has_include(<sys/nope.h>)
#define COMPOSED 1
#else
#define COMPOSED 0
#endif

/* Both branches of an #if that is not taken are skipped, operator and all:
   an unresolved __has_include inside a dead branch is not an error. */
#if 0
#if __has_include(<still/not/scanned.h>)
#endif
#endif

int main(void) {
    if (HAVE_STDIO != 1) return 1;
    if (HAVE_BOGUS != 0) return 2;
    if (HAVE_LOCAL != 1) return 3;
    if (HAVE_LOCAL_BOGUS != 0) return 4;
    if (COMPOSED != 1) return 5;

    /* And the header it found really is includable, which is the claim the
       operator makes. */
#if __has_include("has_a/present.h")
#include "has_a/present.h"
#endif
#ifndef PRESENT_H_WAS_READ
    return 6;
#endif

    printf("__has_include OK\n");
    return 0;
}
